lib/ansible/config/manager.py

```diff
@@ -0,0 +1,242 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+# Make coding more python3-ish
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os
+import sys
+import tempfile
+import yaml
+
+from ansible.config.data import ConfigData, Setting
+from ansible.errors import AnsibleOptionsError, AnsibleError
+from ansible.module_utils.six import string_types
+from ansible.module_utils.six.moves import configparser
+from ansible.module_utils._text import to_text, to_bytes, to_native
+from ansible.parsing.quoting import unquote
+from ansible.utils.path import unfrackpath
+from ansible.utils.path import makedirs_safe
+
+
+def resolve_path(path):
+
+    if '{{CWD}}' in path: # allow users to force CWD using 'magic' {{CWD}}
+        path = path.replace('{{CWD}}', os.getcwd())
+
+    return unfrackpath(path, follow=False)
+
+def get_ini_config(p, entries):
+    ''' returns the value of last ini entry found '''
+    value = None
+    if p is not None:
+        for entry in entries:
+            try:
+                value = p.get(entry.get('section','defaults'), entry.get('key',''), raw=True)
+            except:
+                pass
+
+    return value
+
+
+
+class ConfigManager(object):
+
+    def __init__(self, conf_file=None):
+
+        self.data = ConfigData()
+
+        #FIXME: make dynamic?
+        bconfig_def = to_bytes('%s/data/config.yml' % os.path.dirname(__file__))
+        if os.path.exists(bconfig_def):
+            with open(bconfig_def, 'r') as config_def:
+                self.initial_defs = yaml.safe_load(config_def)
+        else:
+            raise AnsibleError("Missing base configuration definition file (bad install?): %s" % to_native(bconfig_def))
+
+        ftype = None
+        if conf_file is None:
+            # set config using ini
+            conf_file = self.find_ini_config_file()
+            ftype = 'ini'
+        else:
+            ext = os.path.splitext(conf_file)[-1]
+            if ext in ('.ini', '.cfg'):
+                ftype = 'ini'
+            elif ext in ('.yaml', '.yml'):
+                ftype = 'yaml'
+            else:
+                raise AnsibleOptionsError("Unsupported configuration file extension: \n{0}".format(ext))
+
+        self.parse_config(conf_file, ftype)
+
+    def parse_config(self, cfile, ftype):
+        # TODO: take list of files with merge/nomerge
+
+        parser = None
+        if ftype == 'ini':
+            parser = configparser.ConfigParser()
+            try:
+                parser.read(cfile)
+            except configparser.Error as e:
+                raise AnsibleOptionsError("Error reading config file: \n{0}".format(e))
+        elif ftype == 'yaml':
+            with open(to_bytes(cfile, 'r')) as config_stream:
```

Comment:
Not sure, but looks like the `to_bytes` is for the filename, not the contents of the file.  But adding `rb` would probably be beneficial.